REGRESSION(264515@main): m_player WeakPtr is accessed from non-main threads

jyavenard opened the pull request:
#### c234d2219797395a7e23f45a681da9878c44f99a
<pre>
REGRESSION(264515@main): m_player WeakPtr is accessed from non-main threads
<a href="https://bugs.webkit.org/show_bug.cgi?id=258022">https://bugs.webkit.org/show_bug.cgi?id=258022</a>
rdar://110817538

Reviewed by Youenn Fablet.

Tighten WeakPtr assertion that it can only be accessed on the GC thread
if it was created on the main thread.
We apply the same requirement to operator-&gt; and operator* for consistency.

When the GC thread is running, the main thread is suspended which makes
it safe to use with main thread objects.

* Source/WTF/wtf/WeakPtr.h:
(WTF::WeakPtr::get const):
(WTF::WeakPtr::operator-&gt; const):
(WTF::WeakPtr::operator* const):
(WTF::WeakPtr::canSafelyBeUsed const):

Canonical link: <a href="https://commits.webkit.org/265233@main">https://commits.webkit.org/265233@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7b5e1c6b44ea69712631de6cbce182d734e8432a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10282 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10520 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10786 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11929 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9896 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12514 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10470 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12847 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10440 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11173 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8665 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12315 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8481 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9314 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16597 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/8691 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9579 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9465 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12719 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/9760 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9914 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8046 "2 failures") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/10416 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9073 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2694 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2477 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13323 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/10699 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9753 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2640 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WTF/wtf/WeakPtr.h:
```diff
@@ -123,7 +123,7 @@ template<typename T, typename WeakPtrImpl> class WeakPtr {
     T* get() const
     {
         // FIXME: Our GC threads currently need to get opaque pointers from WeakPtrs and have to be special-cased.
```
Should move this comment into your new function.

jyavenard replied:
Done

mcatanzaro reviewed (changes requested):
Your commit message is not correct:

> Tighten WeakPtr so that it can only be accessed on the GC thread if it
was created on the main thread.

But that doesn't match your code changes.

youennf reviewed (approved):
The assert is not handing the case of weak pointers created in workers and checked in GC thread but EWS Mac debug bot will tell us whether this is fine.
r=me once bots are happy.

mcatanzaro reviewed (commented):
.

jyavenard committed:
REGRESSION(264515@main): m_player WeakPtr is accessed from non-main threads
https://bugs.webkit.org/show_bug.cgi?id=258022
rdar://110817538

Reviewed by Youenn Fablet.

Tighten WeakPtr assertion that it can only be accessed on the GC thread
if it was created on the main thread.
We apply the same requirement to operator-> and operator* for consistency.

When the GC thread is running, the main thread is suspended which makes
it safe to use with main thread objects.

* Source/WTF/wtf/WeakPtr.h:
(WTF::WeakPtr::get const):
(WTF::WeakPtr::operator-> const):
(WTF::WeakPtr::operator* const):
(WTF::WeakPtr::canSafelyBeUsed const):

Canonical link: https://commits.webkit.org/265233@main
